Simple version of the iOS back gesture.

mpcomplete opened the pull request:
Doesn't do any of the fancy effects. Just lets the user control the
back transition by sliding from the left, like a drawer. Hero
transitions are disabled during the gesture.

BUG=https://github.com/flutter/flutter/issues/4817

Hixie commented on packages/flutter/lib/src/animation/animations.dart:
```diff
@@ -533,3 +533,58 @@ class TrainHoppingAnimation extends Animation<double>
   }
 }
+
+/// A simple interface for combining multiple Animations. Subclasses need only
```
strike "simple" (we let the user decide what's simple and what's not)

mpcomplete replied:
Done

mpcomplete commented:
There's a bit of a delay (at least in Slow Mode) when first sliding the previous page from the left, likely due to rebuilding it. We might need to cache a snapshot of the routes to speed up the transition.

Hixie commented on packages/flutter/lib/src/animation/animations.dart:
```diff
@@ -533,3 +533,58 @@ class TrainHoppingAnimation extends Animation<double>
+    first.removeListener(notifyListeners);
+    first.removeStatusListener(notifyStatusListeners);
+    next.removeListener(notifyListeners);
+    next.removeStatusListener(notifyStatusListeners);
```
this will send redundant status messages. For example, if next.status isn't dismissed, then as first toggles direction we'll keep firing notifications but the status won't have changed.

mpcomplete replied:
Done

Hixie commented on packages/flutter/lib/src/material/page.dart:
```diff
@@ -5,12 +5,19 @@
   end: FractionalOffset.topLeft
 );
 
+final FractionalOffsetTween _kIosPageTransitionTween = new FractionalOffsetTween(
```
for privates, use "Cupertino" and "MountainView" instead of "iOS" and "Android" so that the capitalisation of "iOS" doesn't cause problems (like it does here, with "Ios")

mpcomplete replied:
Done

Hixie commented on packages/flutter/lib/src/material/page.dart:
```diff
@@ -5,12 +5,19 @@
+import 'material.dart';
+import 'theme.dart';
 
 final FractionalOffsetTween _kMaterialPageTransitionTween = new FractionalOffsetTween(
```
all the mentions of "Material" here should be "Android". Both the iOS and Android transitions are "material", they're just adapted to the local host platform.

mpcomplete replied:
Done

Hixie commented on packages/flutter/lib/src/material/page.dart:
```diff
@@ -36,6 +43,59 @@ class _MaterialPageTransition extends AnimatedWidget {
+
+  @override
+  Widget build(BuildContext context) {
+    // TODO(ianh): tell the transform to be un-transformed for hit testing
```
mention in this TODO that this should only be done during the animation, not while the finger is driving it

mpcomplete replied:
Done

Hixie commented on packages/flutter/lib/src/material/page.dart:
```diff
@@ -36,6 +43,59 @@ class _MaterialPageTransition extends AnimatedWidget {
+  }
+}
+
+class AnimationSum extends CompoundAnimation<double> {
```
This is a mean, not a sum.

mpcomplete replied:
Done

Hixie commented on packages/flutter/lib/src/material/page.dart:
```diff
@@ -64,9 +124,6 @@ class MaterialPageRoute<T> extends PageRoute<T> {
   Color get barrierColor => null;
 
   @override
-  bool canTransitionFrom(TransitionRoute<dynamic> nextRoute) => false;
```
?

If I'm not mistaken, this will cause crazy things to happen like heroes to animate into dialogs.

mpcomplete replied:
Is there a way to see the effect of this? If I don't remove this, the previous (leftmost) route won't animate during a transition.

Hixie replied:
You want to return true in the specific case of transitioning from one route that supports this to another that supports this, but you don't want to return true across the board. To see why, put a hero on a page then show a dialog with a matching hero. If I'm not mistaken, if this returns true, it'll try to animate from one to the other. Or indeed, put your iOS-style route before a dialog route, same thing without heroes.

mpcomplete replied:
So this should be:
  bool canTransitionFrom(TransitionRoute<dynamic> nextRoute) {
    return nextRoute is MaterialPageRoute<dynamic>;
  }
?

Hixie replied:
Yes, at least, in the case where it's going to do the cupertino thing. Maybe not on Android?

mpcomplete replied:
Is there a way to query the platform without access to the BuildContext or Theme?

Hixie replied:
Not a good way. You could ask it to be passed in as an argument to the route, I guess?

Hixie replied:
Or just don't worry about not doing it for android?

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -686,6 +688,33 @@ class ScaffoldState extends State<Scaffold> {
     );
   }
 
+  // IOS-specific back gesture.
```
iOS

mpcomplete replied:
Done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -770,6 +799,23 @@ class ScaffoldState extends State<Scaffold> {
+            key: _backGestureKey,
+            onHorizontalDragStart: _handleDragStart,
+            onHorizontalDragUpdate: _handleDragUpdate,
+            onHorizontalDragEnd: _handleDragEnd,
```
also handle onHorizontalDragCancel, otherwise if the drag is canceled the navigator ends up in a zombie state

mpcomplete replied:
Done. But how does one trigger a drag cancel? I can only get it to call End.

Hixie replied:
You can get a cancel if e.g. the system pops up a dialog.

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -686,6 +688,33 @@ class ScaffoldState extends State<Scaffold> {
+  }
+
+  void _handleDragUpdate(DragUpdateDetails details) {
+    Navigator.of(context).currentRouteController?.value -= details.primaryDelta / 300.0;
```
what is this 300.0 constant?

mpcomplete replied:
screensize for the lazy

Hixie commented on packages/flutter/lib/src/widgets/heroes.dart:
```diff
@@ -467,8 +467,21 @@ class HeroController extends NavigatorObserver {
     }
   }
 
+  // Suspend Hero animations while a user gesture is controlling the navigation.
```
if i'm reading this right it's not suspending active animations, it's preventing new ones from starting. Is that what you meant? If so I'd call this _questsEnabled rather than _suspendQuest (and flip the meaning of the boolean)

mpcomplete replied:
Done.

Hixie commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -10,6 +10,7 @@ import 'binding.dart';
 import 'focus.dart';
 import 'framework.dart';
 import 'overlay.dart';
+import 'pages.dart';
```
this is a layering violation. The API is designed so that Navigator doesn't know about Routes.

mpcomplete replied:
Done

Hixie commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -460,6 +468,18 @@ class NavigatorState extends State<Navigator> {
+  void didStartUserGesture() { config.observer?.didStartUserGesture(); }
+  void didStopUserGesture() { config.observer?.didStopUserGesture(); }
+
+  AnimationController get currentRouteController {
```
This kind of pattern, where two objects can end up both controlling the same animation controller, has historically been a big source of bugs.

Hixie commented on packages/flutter/lib/src/widgets/routes.dart:
```diff
@@ -109,6 +109,8 @@ abstract class TransitionRoute<T> extends OverlayRoute<T> {
   Animation<double> get animation => _animation;
   Animation<double> _animation;
+
+  AnimationController get controller => _controller;
```
I don't think we should expose the controller.

Hixie commented on packages/flutter/lib/src/widgets/navigator.dart:
```diff
@@ -460,6 +468,18 @@ class NavigatorState extends State<Navigator> {
     return _history.length > 1 || _history[0].willHandlePopInternally;
   }
 
+  void didStartUserGesture() { config.observer?.didStartUserGesture(); }
```
By convention, foo.didDoBar() methods are called by the owner of the foo object to notify it that it did bar. (So on the observer class it makes sense.)
This API is more about notifying a peer object (the navigator) that something happened, and asking it to do something about it. So I'd go with something more like "startGesture".

Hixie replied:
Having said that, the general pattern of having paired calls to maintain state is pretty error prone, so maybe we should find a different approach.

mpcomplete replied:
Changed this to a helper interface vended by the Route and Navigator, as we discussed.

Hixie commented:
The overall approach here is good but I think the API itself should be tweaked a bit.

I think rather than just exposing the controller directly, and having start/stop methods for the user gesture, I would hand the navigator (or maybe the route) the pointer directly, and have the navigator (or route) then do the gesture recognition. This would remove the method pair on the API, and would mean the owner of the animation controller could directly control the animation, instead of handing it out to another object and just trusting that that other object doesn't do bad things with it.

Hixie commented:
What are the fancy effects that this doesn't do?

mpcomplete commented:
> I would hand the navigator (or maybe the route) the pointer directly

Which pointer?

> What are the fancy effects that this doesn't do?

Probably among other things, the native ios gesture keeps the app bar and back button in tact, while slowly animating and fading in the previous route's title.

Hixie commented:
> > I would hand the navigator (or maybe the route) the pointer directly
> Which pointer?

Basically, I'd replace the GestureDetector with a PointerListener and pass any pointer down events it sees to the route, which would then pass them to its own local drag gesture recognizer.

> Probably among other things, the native ios gesture keeps the app bar and back button in tact, while slowly animating and fading in the previous route's title.

Ah, interesting. For us, those are hero animations. So that might be incompatible with preventing heroes.

Maybe we should reach out to the material UX people and see what they want us to do when we're doing the back gesture on iOS between two pages whose transition normally includes heroes. @sethladd?

Hixie commented:
Filed https://github.com/flutter/flutter/issues/5535 to track this last issue.

mpcomplete commented:
OK, I've addressed all of your comments except for the one about removing MaterialPageRoute.canTransitionFrom. I'm not sure what to do there yet.

Hixie commented on packages/flutter/lib/src/animation/animations.dart:
```diff
@@ -533,3 +533,66 @@ class TrainHoppingAnimation extends Animation<double>
+/// An interface for combining multiple Animations. Subclasses need only
+/// implement the `value` getter to control how the child animations are
+/// combined. Can be chained to combine more than 2 animations.
+/// Example: to create an animation that is the sum of two others, subclass
```
blank line before the example.
introduce the example with "For example, to create...".
Put the "value = ..." stuff in back ticks.

Hixie commented on packages/flutter/lib/src/animation/animations.dart:
```diff
@@ -533,3 +533,66 @@ class TrainHoppingAnimation extends Animation<double>
+  }
+
+  AnimationStatus _lastStatus;
+  void _maybeNotifyStatusListeners(AnimationStatus _) {
```
avoid `_` as an argument name, to ease copy/pasting later.

mpcomplete replied:
What copy/pasting? I'm using _ because it's an unused parameter.

Hixie replied:
Say I want to write a status listener some day. I happen to be near this code. If you have the argument with a name, I can just swipe it and reuse it and it's self-documenting. If you use `_`, then I either have to find another one, or have to look it up in the docs to find out what that argument actually is.

mpcomplete replied:
I don't think naming the parameter `status` is much more illuminating, given the type name. And in any case, I think code clarity here is more important, to show that the parameter is unused.

Hixie replied:
It's a little confusing because this argument is normally called status, and there is an identifier with that name in scope here that you do use, it's just not the argument. I dunno what to suggest. I worry that one day I'll be going through the code and will change this to "status" reflexively so that the code is more self-documenting, and will thus break the method (because now it'll mean something else).

Maybe what that tells us is that what we really need here is a test of this class, in particular a test that verifies that this function works as expected. :-)

mpcomplete replied:
Good point that it may be confusing that I reference a `status` var here. I'll change it to `this.status` to be extra obvious.

Hixie commented on packages/flutter/lib/src/animation/animations.dart:
```diff
@@ -533,3 +533,66 @@ class TrainHoppingAnimation extends Animation<double>
+      notifyStatusListeners(status);
+    }
+  }
+}
```
should we do the same with the value, to avoid sending two callbacks per frame when both are animating?

mpcomplete replied:
AFAICT we can't know when they're both animating, since either could be paused mid-animation.

Hixie replied:
Sure. I'm saying, check what value you would return, and only fire the notifications if the value itself changed. That way when both _are_ animating, you only fire one notification here, rather than two, with the second not having actually changed.

mpcomplete replied:
Done

Hixie commented on packages/flutter/lib/src/material/page.dart:
```diff
@@ -5,20 +5,22 @@
-  begin: FractionalOffset.bottomLeft,
-  end: FractionalOffset.topLeft
-);
+class _AndroidPageTransition extends AnimatedWidget {
```
_MountainViewPageTransition

mpcomplete replied:
Done

Hixie commented on packages/flutter/lib/src/widgets/heroes.dart:
```diff
@@ -467,8 +467,21 @@ class HeroController extends NavigatorObserver {
+
   void _checkForHeroQuest() {
-    if (_from != null && _to != null && _from != _to) {
+    if (_from != null && _to != null && _from != _to && _questsEnabled) {
```
fwiw, i think we'll find that what we actually want is to categorise heroes into those that want to participate during one of these gestures and ones that don't. but that can wait.

mpcomplete replied:
Agreed.

Hixie commented on packages/flutter/lib/src/material/page.dart:
```diff
@@ -36,6 +38,99 @@ class _MaterialPageTransition extends AnimatedWidget {
+    } else {
+      controller.forward();
+    }
+    dispose();
```
should this call super.dragEnd and have the superclass call dispose?

mpcomplete replied:
Done

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -686,6 +688,36 @@ class ScaffoldState extends State<Scaffold> {
+
+  void _handleDragUpdate(DragUpdateDetails details) {
+    _backGestureController?.dragUpdate(
+      details.primaryDelta / MediaQuery.of(context).size.width
```
i think this should use `findRenderObject().size.width` rather than the screen width.

mpcomplete replied:
Done.

mpcomplete commented:
Crap. I just noticed that using `forwardAnimation` in the transition breaks Hero animations when `pop()`ing. Investigating.

mpcomplete commented:
OK, I pushed a new version addressing the Hero ugliness. We now only use iOS-style transitions when doing the user gesture. The BackGestureController also lives until the transition is complete, so we can disable Heroes for the full transition.

Hixie commented on packages/flutter/lib/src/material/page.dart:
```diff
@@ -5,20 +5,23 @@
-  begin: FractionalOffset.bottomLeft,
-  end: FractionalOffset.topLeft
-);
+// Used for android and fuschia.
```
Android and Fuchsia

Hixie commented on packages/flutter/lib/src/material/page.dart:
```diff
@@ -36,6 +39,108 @@ class _MaterialPageTransition extends AnimatedWidget {
   }
 }
 
+// Used for iOS.
```
Maybe add that it's used for iOS only when driven by touch?

Hixie commented on packages/flutter/lib/src/material/page.dart:
```diff
@@ -36,6 +39,108 @@ class _MaterialPageTransition extends AnimatedWidget {
+    controller?.addStatusListener(onStatusChanged);
+  }
+
+  void onStatusChanged(AnimationStatus status) {
```
this should be called handleStatusChanged. The convention is onFoo is the property, handleFoo is the closure.

Hixie commented on packages/flutter/lib/src/material/page.dart:
```diff
@@ -36,6 +39,108 @@ class _MaterialPageTransition extends AnimatedWidget {
+      dispose();
+  }
+
+  AnimationController controller;
```
i'd move these to just after the constructor, since that way you see them from the constructor

Hixie commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -686,6 +688,35 @@ class ScaffoldState extends State<Scaffold> {
+
+  bool _shouldHandleBackGesture() {
+    return Navigator.canPop(context);
+    // return Theme.of(context).platform == TargetPlatform.iOS && Navigator.canPop(context);
```
don't check in commented out code

Hixie commented:
![LGTM, but travis is angry.](https://cloud.githubusercontent.com/assets/551196/16323131/36570fa8-395d-11e6-9822-7a0aac54e3b5.png)

mpcomplete committed:
Simple version of the iOS back gesture.

Doesn't do any of the fancy effects. Just lets the user control the
back transition by sliding from the left, like a drawer. Hero
transitions are disabled during the gesture.

BUG=https://github.com/flutter/flutter/issues/4817
